src/main/scala/org/camunda/feel/impl/parser/FeelParser.scala

```diff
@@ -43,92 +56,133 @@ object FeelParser extends JavaTokenParsers {
+    P(
+      StringIn(
+        "date and time",
+        "years and months duration",
+        "string length",
+        "upper case",
+        "lower case",
+        "substring before",
+        "substring after",
+        "starts with",
+        "ends with",
+        "list contains",
+        "insert before",
+        "index of",
+        "distinct values",
+        "get entries",
+        "get value",
+        "is defined",
+        "day of year",
+        "day of week",
+        "month of year",
+        "week of year"
+      )
+    )
 
   // list of built-in function parameter names with whitespaces
   // -- other names match the 'parameter name' pattern
-  private def builtinFunctionParameterNames[_: P]: P[String] =
-    P[String](
-      StringIn("start position", "grouping separator", "decimal separator").!)
+  private def builtinFunctionParameterNames[_: P] =
+    P(StringIn("start position", "grouping separator", "decimal separator"))
 
-  private def identifier[_: P]: P[String] = P(!(reservedWord ~~ !ident) ~ ident)
+  // an identifier which is not a reserved word. however, it can contain a reserved word.
+  private def identifier[_: P] = P(reservedWord.? ~~ javaLikeIdentifier)
+
+  private def javaLikeIdentifier[_: P] =
+    P(
+      CharPred(Character.isJavaIdentifierStart) ~~ CharsWhile(
+        Character.isJavaIdentifierPart,
+        0)
+    )
 
-  // Java-like string literal: ("\""+"""([^"\x00-\x1F\x7F\\]|\\[\\'"bfnrt]|\\u[a-fA-F0-9]{4})*"""+"\"")
-  // modification: allow '\'
   private def stringLiteralWithQuotes[_: P]: P[String] =
-    //P[String](""""([^"\x00-\x1F\x7F]|\\u[a-fA-F0-9]{4})*"""".r.map(_.replaceAll("\"", "")))
-    P[String]("\"" ~~ CharsWhile(_ != '"', 0).! ~~ "\"")
+    P("\"" ~~ (!"\"" ~~ AnyChar.!).repX ~~ "\"")
+      .map(_.mkString)
 
   // 1 a)
-  private def expression[_: P]: P[Exp] = P[Exp](textualExpression)
+  private def expression[_: P]: P[Exp] = P(textualExpression)
+
   // 1 b)
   private def expression10[_: P] = P(boxedExpression)
 
   // 3
   private def textualExpressions[_: P]: P[ConstList] =
-    P[ConstList](
-      textualExpression.rep(1, sep = ",").map(s => ConstList(s.toList)))
+    P(textualExpression.rep(1, sep = ",").map(s => ConstList(s.toList)))
 
   // 2 a)
   private def textualExpression[_: P]: P[Exp] =
-    P[Exp](
-      functionDefinition | forExpression | ifExpression | quantifiedExpression | expression2)
+    P(functionDefinition | forExpression | ifExpression | quantifiedExpression | expression2)
+
   // 2b)
-  private def expression2[_: P] = P(disjunction)
+  private def expression2[_: P]: P[Exp] = P(disjunction)
+
   // 2 c)
-  private def expression3[_: P] = P(conjunction)
+  private def expression3[_: P]: P[Exp] = P(conjunction)
+
   // 2 d)
   private def expression4[_: P]: P[Exp] =
-    P[Exp](expression5.flatMap(optionalComparison))
+    P(expression5.flatMap(x =>
+      comparison(x).?.map(_.fold(x)(comparisonExpr => comparisonExpr))))
```

Comment:
Yes, this part is a bit tricky :smile:

`flatMap()` is an optimization. It could be also written as `expression5 ~ comparison | expression5`. With the `flatMap()`, we need to parse `expression5` only once.

`_.fold(x)(...)` is applied on the `Option` of `comparison(x).?`. It returns `x` if no comparison is present. Otherwise, it returns the comparison. See the Scala docs [here](https://www.scala-lang.org/api/current/scala/Option.html#fold[B](ifEmpty:=%3EB)(f:A=%3EB):B).

Reply:
Thanks, the docs really cleared it up for me.